DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,172 @@
+
+## Rationale
+Writing the same enum type names over and over again (e.g. exhaustive enum switch-cases)
+involves the unnecessary, often gratuitous repetition of the enum member's type name.
+
+The solution used by many other modern language is simple: permit the omission
+of the enum member's type name when it can be inferred from its context.
+Dlang having this shortcut would be equally beneficial, with few drawbacks.
+
+## Prior Work
+Implementation of this feature in other languages:
+- [Swift](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html)
+- [Ziglang](https://ziglang.org/documentation/master/#Enum-Literals)
+- [Odin](https://odin-lang.org/docs/overview/#implicit-selector-expression)
+
+Java allows omitting an enum type in switch-case statements: [https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)
+
+A suggestion for a similar feature in Rust, but with a different approach: [https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851](https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851)
+
+## Description
+Due to `.enumMember` being the same syntax as the [module scope operator](https://dlang.org/spec/module.html#module_scope_operators),
+I propose that ETI should use the syntax `$enumMember`. I'm open to discussing alternatives,
+but I believe that my proposed syntax represents the best compromise between convenience and compatibility.
+
+#### 1. Assignment statements.
+Assigning to a variable with a known type should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void main(){
+    A    myA1 = $b; //myA1 = A.b
+    auto myA2 = $b; //error, we don't know the type of "$b"!
+    
+    B myB;
+    myB.one = $c; //myB.one = A.c
+    myB.two = $d; //myB.two = A.d
+}
+```
+
+#### 2. Return statements.
+If a function has an explicit enum return type, its return statement(s)
+should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+A myFunc(){
+    return $c; //returns A.c
+}
+
+auto myBrokenFunc(){
+    return $c; //error, we don't know the type of "$c"!
+}
+```
+
+#### 3. Parameters.
+Most parameters require explicit typing beforehand, and thus should
+always allow ETI. The exception is `alias` template parameters, which
+cannot use ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void myFunc(A param){}
+void myTempFunc(T)(T param){}
+
+void main(){
+    B    myB1 = {one:$a, two:$b};
+    auto myB2 = B($a, $b);
+    
+    myFunc($a);
+    myTempFunc!A($a);
+    myTempFunc($a); //error, can't infer a type to instantiate template with from "$a"
+}
+```
+
+#### 4. Switch-case statements.
+Switch-case statements should have a special syntax that works the same way as `switch(myVar) with(typeof(myVar))`.
+```d
+enum WordLetterOfTheDay{ a,b,c,d/*...*/; }
+
+void main(){
+    auto letterToday = WordLetterOfTheDay.b;
+    
+    import std.stdio;
+    switch(letterToday){
+        case a:
```

Comment:
By "already in scope", do you mean `switch(letterToday) with(WordLetterOfTheDay){`?
If so, I believe that this syntax should be ignored for enum types. Part of my goal with this DIP is that it should not introduce any breaking changes.

Reply:
Say you have `int a;` or `WordLetterOfTheDay b;` declared before the switch. Those could conflict with the enum member names.

Reply:
Currently the dip implies those declarations would be ignored, but that could potentially break existing code.

Reply:
> Say you have `int a;` or `WordLetterOfTheDay b;` declared before the switch. Those could conflict with the enum member names.

Currently this exact case would cause an error (as of dmd 2.100.0):
```d
enum A{ a,b,c,d }
int a=0;

void main(){
	A myA = A.a;
	switch(myA){
		case a:
			break;
		default:
	}
}
```
```
source/app.d(9,8): Error: cannot implicitly convert expression `a` of type `int` to `A`
```
I suppose this DIP could potentially cause broken old code to compile if it was written this way, but that's not a breaking change.

Here's an actual breaking change that I didn't realise could be caused by this:
```d
enum A{ a,b,c,d }
alias a=A.b;

void main(){
	A f = A.a;
	switch(f){
		case a: //is this A.a or A.b??
			break;
		default:
	}
}
```
What if I revise the DIP so that switch-case statements with enums require the `$enumMember` syntax, but also introduce a shorthand for `switch(myVar) with(typeof(myVar))`: `switch with(myVar)`

This would be useful for more than just enums, of course. My main concern is that I may have overlooked an issue with using `with` in this way, as I've never actually used it myself.

Another option is to prioritise non-enum members here unless there's a `$` prefix, but I feel that this would be obtuse and cause very hard-to-diagnose bugs.

I'll revise the DIP in the next few days with this issue in mind, thank you for bringing it to my attention.

Reply:
> so that switch-case statements with enums require the $enumMember syntax, 

Yes that's the most consistent and avoids ambiguity.

> but also introduce a shorthand for switch(myVar) with(typeof(myVar)): switch with(myVar)

That would avoid breaking code, though I think having to write `$member` for each case is trivial and improves readability in a long switch statement.

Reply:
> That would avoid breaking code, though I think having to write `$member` for each case is trivial and improves readability in a long switch statement.

True. It's still more effort per case than with the old `switch(myVar) with(typeof(myVar))`, which presents a strange old-school/new-school way of writing short enum switch-cases. I'll have to sleep on it, but I think the `switch with` syntax could be really useful too, unless I'm missing something bad. :thinking: